slots (WIP)

tjhance opened the pull request:
There are still a couple of tests failing (including the one with the recursive GC call that I talked about on Gitter... I'm still not sure if that's partially the fault of this diff or not), but I'll go ahead and put it up here. Also, this ended up being one of those diffs where I fix one thing and that leads to fixing another thing...

Anyway, here is what is happening in this diff:
- In `typeNew`, I add support for `__slots__`. This adds a member descriptor for each name in the slots array. Also, if `__dict__` is in the slots, then it adds a dict (well, really an attrs, since it's pyston).
- Hopefully this should reduce allocations since most objects with __slots__ won't have to allocate attrs.
- To garbage collect the slots, we need to keep track of the slot locations in the type object. CPython lists `PyMemberDef`s at the end of a type object. We don't need the full `PyMemberDef` struct (as far as I know--maybe the c api requires it though), but for now, I just put the slot offsets in a variable array in the type object.
- In the process of writing the above, logic, I stumbled across a tangential problem with the way we add a dict (or in our case, attrs) to a class when subclassing a class that is variable-size. This is tricky, since you have to leave the base layout alone, but that means you don't have a fixed offset to put the dict afterwards. CPython solves this by allowing `tp_dictoffset` to be negative, so you can offset it from the *end* of the object. From the [https://docs.python.org/2/c-api/typeobj.html#c.PyTypeObject.tp_dictoffset](python docs):

> If the value is less than zero, it specifies the offset from the end of the instance structure. A negative offset is more expensive to use, and should only be used when the instance structure contains a variable-length part. This is used for example to add an instance variable dictionary to subtypes of str or tuple. Note that the tp_basicsize field should account for the dictionary added to the end in that case, even though the dictionary is not included in the basic object layout.

In the process of supporting this, I had to make sure that string and tuple subclassed `VarBox` and set the `ob_size` field. It's a bit messy and I think more can be done here to clean things up, but I didn't want to dig too far on this since, again, it's kind of tangential.

kmod commented on src/runtime/objmodel.cpp:
```diff
@@ -607,8 +624,11 @@ Box* Box::getattr(const std::string& attr, GetattrRewriteArgs* rewrite_args) {
-            if (!rewrite_args->obj_hcls_guarded)
+            if (!rewrite_args->obj_hcls_guarded) {
+                if (cls->attrs_offset < 0)
+                    REWRITE_ABORTED("");
```
`REWRITE_ABORTED` doesn't actually abort the rewrite, but is just supposed to be a way to figure out where the abort happened; I think you still need to set rewrite_args to NULL.

kmod commented:
I chatted a bit offline with chris and looked into the GC issue, and it looks like this is an existing problem, but it basically shouldn't ever activate.  Looks like we're adding weaklist pointers on variable-sized classes that shouldn't get them -- you probably need to do something with the code in `BoxedClass::BoxedClass` that sets weaklistoffset, now that that's being handled in typeNew.

tjhance commented:
OK, with that I think I got the kinks out.

tjhance commented:
OK, I did a bit of cleaning up with declaring the variable-size class stuff.

tjhance commented:
got this to pass the build

kmod commented on src/runtime/objmodel.cpp:
```diff
@@ -3900,6 +3945,52 @@ static void assertInitNone(Box* obj) {
+            raiseExcHelper(TypeError, "__slots__ must be identifiers");
+        }
+    }
+}
```
If these two are the same except for exception handling, we could maybe just make one call the other?  It's not great from a perf standpoint but I don't think that'll be a big deal here.

also, it's nice to add a `noexcept` specifier to functions that throw CAPI exceptions (`valid_identifier` version) both as documentation and to force the code to terminate if an exception does try to propagate out of it.

tjhance replied:
Actually it looks like I copied one from cpython so I could look at it as I re-wrote it in pyston exception-style, and then forgot to delete the cpython one.

kmod commented on test/tests/python_slots_test.py:
```diff
@@ -0,0 +1,68 @@
+    class G(C, D):
+        pass
+except TypeError as e:
+    print e.message
```
Could we add a test that adds something like 1000 slot names?  just as an extra check that the memory accesses are safe, since for example class objects are quite large and there's a lot of room in there to overwrite unused fields.

tjhance replied:
yeah more tests is good :) and I should probably also write a test case for a class with a `__metaclass__`

kmod commented on src/core/types.h:
```diff
@@ -434,6 +434,7 @@ class Box {
     // Add a no-op constructor to make sure that we don't zero-initialize cls
     Box() {}
 
+    void* operator new(size_t size, BoxedClass* cls, size_t nitems) __attribute__((visibility("default")));
```
could/should this constructor live in BoxVar?

tjhance replied:
yeah I think so. I didn't really notice that these were here... who uses them? I had assumed that all subclasses of `Box` needed to specify their own `operator new`

kmod replied:
DEFAULT_CLASS is basically just a wrapper around `Box::operator new()` that adds a default argument, and C++ classes are free to not use it and just inherit the base version (not sure of any examples that do this).  I think it should be safe to move this to BoxVar and then change DEFAULT_CLASS_VAR to call `BoxVar::operator new`?

tjhance replied:
oh I just realized you were commenting on a line I had added in this diff. >_< I'm so silly

tjhance commented:
unfortunately I left my laptop at the office today -__- I'll get to your comments tomorrow.

tjhance commented:
@kmod I implemented all your suggestions

kmod commented:
hmm so reading this again, I think there's a slight issue -- requesting `strlen + 1` items as ob_size is different from CPython, which doesn't count the null terminator as part of nitems and says that ob_size is `strlen`.  ie for the empty string, we will say `ob_size == 1`, but cpython will say `ob_size == 0`.  I *think* we've caught all cases where people don't use `PyString_GET_SIZE` to get the size of a string, but it just shows that people definitely assume sometimes that ob_size = number of characters not including null terminator.

It looks like the way CPython handles this is by doing setting `tp_basicsize = sizeof(PyString_Object) + 1`.  (Well, actually they include a `char ob_sval[1]` array, but that ends up adding 4 bytes due to padding, so they do `tp_basicsize = offsetof(PyString_Object, ob_sval) + 1`.)  Could we do something like this?

Also, not necessary here, but I guess CPython's implementation shows that the data always lives at a constant offset and that we could use an array to reference it, instead of an extra pointer field that points inside the data structure.  Not necessary for this PR, but would be a bonus to do that too :)

kmod commented on src/core/types.h:
```diff
@@ -517,12 +521,57 @@ extern "C" PyObject* PystonType_GenericAlloc(BoxedClass* cls, Py_ssize_t nitems)
 // CPython C API compatibility class:
 class BoxVar : public Box {
 public:
     Py_ssize_t ob_size;
```
Can we add a note that this field gets initialized by `operator new` and not in the constructor?

kmod commented on src/core/types.h:
```diff
@@ -517,12 +521,57 @@ extern "C" PyObject* PystonType_GenericAlloc(BoxedClass* cls, Py_ssize_t nitems)
     Py_ssize_t ob_size;
 
+    BoxVar() {}
     BoxVar(Py_ssize_t ob_size) : ob_size(ob_size) {}
```
Can we either kill this constructor or change it to just assert that the size we passed in matches the size that was already set by operator new?

kmod commented on from_cpython/Include/object.h:
```diff
@@ -479,6 +479,7 @@ typedef struct _heaptypeobject {
                                       see add_operators() in typeobject.c . */
     PyBufferProcs as_buffer;
     PyObject *ht_name, *ht_slots;
+    Py_ssize_t nslots;
```
CPython uses ob_size for this field; could we do the same so that we don't have to add a new field?

tjhance replied:
yep, you're right

kmod commented on src/runtime/objmodel.cpp:
```diff
@@ -427,14 +418,14 @@ BoxedHeapClass* BoxedHeapClass::create(BoxedClass* metaclass, BoxedClass* base,
-        BoxedHeapClass(base, gc_visit, attrs_offset, weaklist_offset, instance_size, is_user_defined, name);
+                                       BoxedTuple* bases, size_t nslots) {
+    BoxedHeapClass* made = new (metaclass, nslots)
+        BoxedHeapClass(base, gc_visit, attrs_offset, weaklist_offset, instance_size, is_user_defined, name, nslots);
```
I can't comment up on that line (why github??), but the BoxedHeapClass ctor calls the BoxedClass ctor which sets ob_size to 0, which we probably don't want.

kmod commented on src/runtime/objmodel.cpp:
```diff
@@ -4002,28 +4064,165 @@ Box* typeNew(Box* _cls, Box* arg1, Box* arg2, Box** _args) {
+        BoxedHeapClass::SlotOffset* base_slot_offsets = base_heap_cls->slotOffsets();
+        memcpy(&slot_offsets[final_slot_names.size()], base_slot_offsets,
+               base_heap_cls->nslots * sizeof(BoxedHeapClass::SlotOffset));
+    }
```
ok spent some time trying to think about why we're doing this and cpython doesn't.  you should check what I'm saying and then add a comment if it's along the right lines :P

at first I thought CPython was directly constructing the member descriptors into this array.  it turns out that they just use the space to create PyMemberDef objects, which PyType_Ready will turn into normal descriptor objects.  It looks like they want to keep them around for the same reason we do, which is for GC traversing of the instances of the class.  except that they crawl the inheritance hierarchy (I guess you just need to check the main base?).

so are we just copying them all into here so that we can scan objects more quickly?  I think it's a decent tradeoff to make -- it increases the size of classes that inherit from classes with slots, but makes objects faster to visit with the GC.

also one thing I was confused about was I thought we were also adding member descriptors for the base slots, which would be user-visible.  so I think this copying is just for gc purposes?

tjhance replied:
Yeah this seems to be what I was thinking. I wanted to put them all here so that GC scanning would be faster and we wouldn't have to crawl the hierarchy.

Yeah so I wasn't sure if we would be required to actually have the `PyMemberDef`s here in a user-visible way, so I'm sort of betting that that we won't be required to (hence why I did something different) but I'm not totally sure. If I'm wrong, well, we'll have to restructure this.

kmod replied:
well it's not very user visible; definitely not at the Python level, and it seems unlikely that there would be extensions that look at them.

let's just comment out PyHeapType_GET_MEMBERS in from_cpython/Include/object.h, and make a note that this difference shouldn't be user visible and help with performance, and I think we should be good to go.

kmod commented on src/runtime/types.cpp:
```diff
@@ -1958,6 +1981,17 @@ extern "C" PyObject* PyObject_Init(PyObject* op, PyTypeObject* tp) noexcept {
+                *(Box**)((char*)op + slotOffsets[i]) = NULL;
+            }
+        }
+    }
```
I think we already memset all of the data to zero on allocation -- I changed this from a `... = NULL` to `assert(... == NULL)` and it seems to work ok.

tjhance commented:
> Also, not necessary here, but I guess CPython's implementation shows that the data always lives at a constant offset and that we could use an array to reference it, instead of an extra pointer field that points inside the data structure. Not necessary for this PR, but would be a bonus to do that too :)

Yep, I thought of that too. Right now we use the `llvm::StringRef` so I'm not sure how to actually implement this suggestion.

kmod commented:
ok let's save the "convert BoxedString to have a array at a constant offset" part for another PR (I just noticed that your "clean up BoxedTuple" change does this for tuples).  I was thinking that `BoxedString::s` could be a function that constructed the StringRef on demand.

tjhance commented:
@kmod I implemented all your suggestions.

The only thing left is to fix this merge conflict: chris added some stat logging code to the `operator new`, which I moved into a macro, so his stuff got removed in the merge. I'll just need to add his code back into the macro.

tjhance commented:
Okay, I added the allocation stat stuff to the new macro. Assuming the me-ci build passes, this is probably ready.

kmod commented on src/core/types.h:
```diff
@@ -446,10 +446,12 @@ class Box {
 
     llvm::iterator_range<BoxIterator> pyElements();
 
+    size_t getHCAttrsOffset();
```
It looks like we have a couple uses of attrs_offset that should be updated to call this instead -- the ones I see are in objmodel.cpp related to rewriting.  Could you take a look?  You can check out the test I added in #511

kmod replied:
oh shoot I forgot you had handled that by just aborting the rewrite when we run into those cases.  I'll take out that test.

kmod commented:
I merged in the change I just made today, and then merged this into master.  :star2:

there was another change that was made that I'm surprised didn't conflict, but there's now an extra char in BoxedString for `interned_state`.  I think this means that we can now save 7 bytes over the current approach of using `sizeof(BoxedString) + 1` as the tp_basicsize by doing the CPython trick of doing `offsetof(BoxedString, charbuf) + 1`

tjhance commented:
Why does that save 7 bytes?

kmod commented:
I think there's now 7 bytes of padding after `char interned_state`, and `sizeof(BoxedString)` will include that padding, but we could instead use it for storing character data.
